optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/DroolsRuleStructure.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.drools.model.DSL;
+import org.drools.model.DeclarationSource;
+import org.drools.model.RuleItemBuilder;
+import org.drools.model.Variable;
+
+/**
+ * Represents the left-hand side of a Drools rule.
+ */
+public abstract class DroolsRuleStructure {
+
+    private static final AtomicInteger COUNTER = new AtomicInteger();
```

Comment:
This "static" bugs me.
If I build a SolverFactory twice (or build another one first or in parallel concurrently), will I get different ID's for my CS-D implementation and therefore make reproduction harder?
Think optaplanner-benchmark AUTO.

This counter must not be a static, but a shared stated, starting for zero for one CS-D build.

Furthermore. ScoreDirector is single threaded. No need for AtomicInteger. But what about ScoreDirectorFactory? I 'd still say no need, but let's discuss :)

Reply:
The IDs here are used plainly to guarantee variable name uniqueness - they have no other functional use, and therefore I would argue that they're irrelevant for the purposes of reproducibility.

However, I see your point. I'll see if I can somehow guarantee uniqueness without the IDs.